src/type_checkers/statement.cr

```diff
@@ -6,24 +6,31 @@ module Mint
     def check(node : Ast::Statement | Ast::WhereStatement) : Checkable
       type = resolve node.expression
 
-      case target = node.target
-      when Ast::TupleDestructuring
-        if target.parameters.size > 1
-          raise StatementNotTuple, {
-            "node" => node.expression,
-            "got"  => type,
-          } unless type.name == "Tuple"
+      _check_statement_target(node.target, node, type)
 
-          raise StatementTupleMismatch, {
-            "parameters" => type.parameters.size.to_s,
-            "variables"  => target.parameters.size.to_s,
-            "node"       => node.expression,
-            "got"        => type,
-          } if type.parameters.size < target.parameters.size
-        end
+      types[node] = type
+    end
+
+    private def _check_statement_target(target : Ast::TupleDestructuring, node, condition)
+      raise StatementNotTuple, {
+        "node" => node.expression,
+        "got"  => condition,
+      } unless condition.name == "Tuple"
+
+      raise StatementTupleMismatch, {
+        "parameters" => condition.parameters.size.to_s,
+        "variables"  => target.parameters.size.to_s,
+        "node"       => node.expression,
+        "got"        => condition,
+      } if target.parameters.size > condition.parameters.size
+
+      target.parameters.each_with_index do |param, index|
+        subcondition = condition.parameters[index]
+        _check_statement_target(param, node, subcondition)
       end
+    end
 
-      types[node] = type
+    private def _check_statement_target(target, node, condition)
```

Comment:
Is this needed for a `nil` branch?

Reply:
No, there are multiple types that the where statement target could have but the only one of them we validate is tuple destructuring and that check is also recursive. Rather than do type checking if statements, I use an overload to catch and ignore anything that is not a tuple destructuring.